R/modules.R

```diff
@@ -321,11 +321,124 @@ modules <- function(..., label = "root") {
 # printing methods ----
 
 #' @rdname teal_modules
+#' @param what character vector specifying which metadata to display. 
+#'   Possible values: "data_sources", "properties", "ui_args", "server_args", "transformers"
 #' @export
-format.teal_module <- function(x, indent = 0, ...) {
-  paste0(paste(rep(" ", indent), collapse = ""), "+ ", x$label, "\n", collapse = "")
+format.teal_module <- function(x, indent = 0, is_last = FALSE, parent_prefix = "", what = c("data_sources", "properties", "ui_args", "server_args", "transformers"), ...) {
+  empty_text <- "─ none"
+  branch <- if (is_last) "└─" else "├─"
+  current_prefix <- paste0(parent_prefix, branch, " ")
+  content_prefix <- paste0(parent_prefix, if (is_last) "   " else "│  ")
+
+  format_list <- function(lst, empty = empty_text) {
+    if (is.null(lst) || length(lst) == 0) {
+      empty
+    } else {
+      paste(names(lst), collapse = ", ")
+    }
+  }
+
+  bookmarkable <- ifelse(isTRUE(attr(x, "teal_bookmarkable")), "Yes", "No")
+  reportable <- ifelse("reporter" %in% names(formals(x$server)), "Yes", "No")
+
+  transformers <- if (length(x$transformers) > 0) {
+    paste(sapply(x$transformers, function(t) attr(t, "label")), collapse = ", ")
+  } else {
+    empty_text
+  }
+
+  output <- paste0(current_prefix, crayon::bold(x$label), "\n")
+  
+  if ("data_sources" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "├─ ", crayon::blue("Data Sources:"), " ", paste(x$datanames, collapse = ", "), "\n"
+    )
+  }
+  if ("properties" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "├─ ", crayon::yellow("Properties:"), "\n",
+      content_prefix, "│  ├─ Bookmarkable: ", bookmarkable, "\n",
+      content_prefix, "│  └─ Reportable: ", reportable, "\n"
+    )
+  }
+  if ("ui_args" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "├─ ", crayon::green("UI Arguments:"), " ", format_list(x$ui_args), "\n"
+    )
+  }
+  if ("server_args" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "├─ ", crayon::green("Server Arguments:"), " ", format_list(x$server_args), "\n"
+    )
+  }
+  if ("transformers" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "└─ ", crayon::magenta("Transformers:"), " ", transformers, "\n"
+    )
+  }
```

Comment:
Maybe this part can be grouped into a function call? And you could have 3 calls. `if` could be a part of the function.

Reply:
I did not like that idea. It will be a function that serves a very small use case and someone working on this file for the first time will have a good time understanding this over switching files and finding that function.
But, I think it's just personal preference. I'm 60-40 right now in favor of just having this if statements.

Reply:
They look quite the same (if is different, colour of crayon, name used in crayon, and whats formatted) but not quite the same so maybe function will not be that conveninent/useful. Function could be in the same file, so it's not that bad. Let's revisit this place, for potential function usage if we ever get more than 3 repetitions in here